src/main/java/org/opensearch/security/resources/ResourceAccessHandler.java

```diff
@@ -0,0 +1,343 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.security.resources;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import org.opensearch.accesscontrol.resources.EntityType;
+import org.opensearch.accesscontrol.resources.ResourceSharing;
+import org.opensearch.accesscontrol.resources.ShareWith;
+import org.opensearch.accesscontrol.resources.SharedWithScope;
+import org.opensearch.common.util.concurrent.ThreadContext;
+import org.opensearch.security.configuration.AdminDNs;
+import org.opensearch.security.support.ConfigConstants;
+import org.opensearch.security.user.User;
+import org.opensearch.threadpool.ThreadPool;
+
+/**
+ * This class handles resource access permissions for users and roles.
+ * It provides methods to check if a user has permission to access a resource
+ * based on the resource sharing configuration.
+ */
+public class ResourceAccessHandler {
+    private static final Logger LOGGER = LogManager.getLogger(ResourceAccessHandler.class);
+
+    private final ThreadContext threadContext;
+    private final ResourceSharingIndexHandler resourceSharingIndexHandler;
+    private final AdminDNs adminDNs;
+
+    public ResourceAccessHandler(
+        final ThreadPool threadPool,
+        final ResourceSharingIndexHandler resourceSharingIndexHandler,
+        AdminDNs adminDns
+    ) {
+        super();
+        this.threadContext = threadPool.getThreadContext();
+        this.resourceSharingIndexHandler = resourceSharingIndexHandler;
+        this.adminDNs = adminDns;
+    }
+
+    /**
+     * Returns a set of accessible resources for the current user within the specified resource index.
+     *
+     * @param resourceIndex The resource index to check for accessible resources.
+     * @return A set of accessible resource IDs.
+     */
+    public <T> Set<T> getAccessibleResourcesForCurrentUser(String resourceIndex, Class<T> clazz) {
+        if (areArgumentsInvalid(resourceIndex, clazz)) {
+            return Collections.emptySet();
+        }
+        final User user = threadContext.getPersistent(ConfigConstants.OPENDISTRO_SECURITY_USER);
```

Comment:
Why are we putting this persistently here?

Reply:
Is it to access the system index within the handler?

Reply:
This is to get user which is then used when creating a resource or listing resources. Without this the values may be null.